modules/host/network.go

```diff
@@ -85,9 +85,18 @@ func (h *Host) initNetworking(address string) (err error) {
 // threadedHandleConn handles an incoming connection to the host, typically an
 // RPC.
 func (h *Host) threadedHandleConn(conn net.Conn) {
-	// We defer the conn.Close so that the conn is still closed if `Add` returns
-	// an error.
-	defer conn.Close()
+	// Close the conn immediately on host.Close.
+	defer func() {
+		// We defer the conn.Close so that the conn is still closed if `Add` returns
+		// an error.
+		defer conn.Close()
+		err := h.tg.Add()
+		if err != nil {
+			return
+		}
+		defer h.tg.Done()
+		<-h.tg.StopChan()
+	}()
```

Comment:
as for the `Add` and `Done`, they aren't strictly necessary here. As discussed, `Close` can be called multiple times without issue, so the only real risk is the channel read blocking forever. But since `tg.Stop` closes the channel, it is guaranteed to return with `tg.Stop`.